src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Tds/Tokens/DataClassification/DataClassificationTokenParser.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClientX.IO;
+using Microsoft.Data.SqlClient.DataClassification;
+
+namespace Microsoft.Data.SqlClientX.Tds.Tokens.DataClassification
+{
+    internal class DataClassificationTokenParser : TokenParser
+    {
+        public override async ValueTask<Token> ParseAsync(TokenType tokenType, TdsStream tdsStream, bool isAsync, CancellationToken ct)
+        {
+            if (tdsStream.DataClassificationVersion == 0)
+            {
+                throw SQL.ParsingError(ParsingErrorState.DataClassificationNotExpected);
+            }
+
+            // get the labels
+            ushort numLabels = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct).ConfigureAwait(false);
+
+            List<Label> labels = new List<Label>(numLabels);
+            for (ushort i = 0; i < numLabels; i++)
+            {
+                var labelLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+                var label = await tdsStream.TdsReader.ReadStringAsync(labelLen, isAsync, ct).ConfigureAwait(false);
+                var idLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+                var id = await tdsStream.TdsReader.ReadStringAsync(idLen, isAsync, ct).ConfigureAwait(false);
+                labels.Add(new Label(label, id));
+            }
+
+            // get the information types
+            ushort numInformationTypes = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct);
+
+            List<InformationType> informationTypes = new List<InformationType>(numInformationTypes);
```

Comment:
These are publicly exposed as list too, so this will be retained.

Reply:
See previous comment response

Reply:
@benrr101, @cheenamalhotra, I spent sometime on this part and I was thinking if we could optimized the loop by starting the string reads immediately after reading the byte lengths for both label and ID.
This reduces the need for separate await calls and allows the label and ID reading operations to be performed in parallel,
potentially improving performance, especially if these operations are I/O-bound (which they likely are since they involve reading from a stream).

Have not tested, just as an idea, something similar to below, again I might be totally wrong, but thought it might worth sharing that with you as well.
```C#
for (ushort i = 0; i < numLabels; i++)
{
    byte labelLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
    ValueTask<string> labelTask = tdsStream.TdsReader.ReadStringAsync(labelLen, isAsync, ct);

    byte idLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
    ValueTask<string> idTask = tdsStream.TdsReader.ReadStringAsync(idLen, isAsync, ct);

    labels.Add(new Label(await labelTask.ConfigureAwait(false), await idTask.ConfigureAwait(false)));
}
```